File: r/loop_translation.R
library("sets")

naive_loop_tr <- function(A, B, C){

    at <- matrix(A, ncol=ncol(A), nrow=nrow(A))
    bt <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- matrix(C, ncol=ncol(C), nrow=nrow(C))
    cache_scrub()
    start.time <- Sys.time()

    for (i in 1:dim(C)[2]){
        ct[, i] = at %*% bt[, i]
    }
    end.time <- Sys.time()
    return(tuple(as.double(end.time-start.time), ct))
}

recommended_loop_tr <- function(A, B, C){

    at <- matrix(A, ncol=ncol(A), nrow=nrow(A))
    bt <- matrix(B, ncol=ncol(B), nrow=nrow(B))
    ct <- matrix(C, ncol=ncol(C), nrow=nrow(C))
    cache_scrub()
    start.time <- Sys.time()
    ct = at %*% bt
    end.time <- Sys.time()
    return(tuple(as.double(end.time-start.time), ct))
}

loop_translation <- function(b, ...){

    res1 = benchmark("loop_translation_nai", naive_loop_tr, ...)
    res2 = benchmark("loop_translation_rec", recommended_loop_tr, ...)
    cat(c("LoopTranslation correctness: ", all.equal(res1, res2), "\n"), sep="")
}

File: python/loop_translation.py
from scipy import linalg
import numpy as np
import logging
import time
import copy
from benchmarker import cache_scrub

logger = logging.getLogger('LoopTranslation')


def loop_translation(b, *args):

    res1 = b.benchmark("loop_translation_nai", naive_loop, *args)
    res2 = b.benchmark("loop_translation_rec", recommended_loop, *args)
    logger.info('LoopTranslation correctness: {}'.format(np.allclose(res1, res2)))


def naive_loop(A, B, C):

    at = copy.deepcopy(A)
    bt = copy.deepcopy(B)
    ct = copy.deepcopy(C)
    cache_scrub()
    start = time.perf_counter()

    for i in range(C.shape[1]):
        ct[:, i] = at @ bt[:, i]
    end = time.perf_counter()
    return end-start, ct


def recommended_loop(A, B, C):

    at = copy.deepcopy(A)
    bt = copy.deepcopy(B)
    ct = copy.deepcopy(C)
    cache_scrub()
    start = time.perf_counter()
    ct = at @ bt
    end = time.perf_counter()
    return end-start, ct
